himalaya/internal/repair/repair.go

```diff
@@ -0,0 +1,365 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package repair
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	cmdfile "github.com/ethersphere/bee-repair/file"
+	"github.com/ethersphere/bee-repair/himalaya/internal/collection/entry"
+	"github.com/ethersphere/bee/pkg/file"
+	"github.com/ethersphere/bee/pkg/file/joiner"
+	"github.com/ethersphere/bee/pkg/file/loadsave"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/manifest"
+	"github.com/ethersphere/bee/pkg/manifest/mantaray"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"io/ioutil"
+)
+
+const (
+	limitMetadataLength = swarm.ChunkSize
+)
+
+// ProgressUpdater is and interface which can be implemented by client to recieve
+// updates from the utility
+type ProgressUpdater interface {
+	Update(string)
+}
+
+// Option is used to supply functional options for the repairer utility
+type Option func(*Repairer)
+
+// WithAPIStore is used to configure the API endpoint for running the utility. This
+// could be locally running bee node or some gateway
+func WithAPIStore(host string, port int, useSSL bool) Option {
+	return func(c *Repairer) {
+		c.store = cmdfile.NewApiStore(host, port, useSSL)
+	}
+}
+
+// WithLogger is used to supply optional logger to see debug messages
+func WithLogger(l logging.Logger) Option {
+	return func(c *Repairer) {
+		c.logger = l
+	}
+}
+
+// WithEncryption is used to enable encryption while creating data
+func WithEncryption(val bool) Option {
+	return func(c *Repairer) {
+		c.encrypt = val
+	}
+}
+
+// WithProgressUpdater is used to provide updater implementation to see updates
+// from utility
+func WithProgressUpdater(upd ProgressUpdater) Option {
+	return func(c *Repairer) {
+		c.updater = upd
+	}
+}
+
+// FileRepair takes in an older file reference and creates a new manifest which contains
+// the file and the metadata. This reference can be then used to query the /bzz endpoint to
+// serve the file
+//
+// Old Entry:
+// collection -> file reference -> file bytes
+//           |
+//           |-> metadata reference -> metadata bytes
+//
+// New Entry:
+// mantaray manifest -> Root Node (\) -> Metadata (index file)
+//                  |
+//                  |-> file entry -> Metadata (Filename, ContentType)
+//                                |
+//                                |-> File reference
+//
+func FileRepair(ctx context.Context, addr swarm.Address, opts ...Option) (swarm.Address, error) {
+	r := newWithOptions(opts...)
+
+	oldEntry, err := r.getOldFileEntry(ctx, addr)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+	r.updater.Update(fmt.Sprintf("Updating reference for file %s", oldEntry.mtdt.Filename))
```

Comment:
just a suggestion, but `Update` could be converted into (or offer on-top-of) `Updatef`, which would call `fmt.Sprintf` with the arguments under the hood

Reply:
Update is to be provided by the client calling the repair package. I would like to keep it simpler from the client-side.